### PR TITLE
Updating assertion check for gradient clipping w/ FSDP

### PR DESCRIPTION
**Pull request by bcui19**
# What does this PR do?
Updates the assertion check for gradient clipping when using FSDP

# What issue(s) does this change relate to?
[CO-1427](https://mosaicml.atlassian.net/jira/software/c/projects/CO/boards/12?modal=detail&selectedIssue=CO-1427&assignee=63127d143778a7aadf1901c7)

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by mvpatel2000 — APPROVED** [review]

**Comment by mvpatel2000** [comment]
Can we add a unit test for this?

**Review by vchiley — APPROVED** [review]

File: tests/algorithms/test_gradient_clipping.py [hunk 1]
```diff
@@ -127,26 +134,44 @@ def test_gradient_clipping_algorithm_with_deepspeed_enabled(
...
 def test_gradient_clipping_algorithm_with_fsdp_enabled(
     monkeypatch,
     clipping_type,
     simple_model_with_grads,
     dummy_state: State,
 ):
+    from torch.distributed.fsdp import FullyShardedDataParallel
+
+    from composer.devices import DeviceGPU
+    from composer.utils import dist
```
Comment (mvpatel2000) [inline tests/algorithms/test_gradient_clipping.py:150]: move to top of file?

File: tests/algorithms/test_gradient_clipping.py [hunk 2]
```diff
@@ -127,26 +134,44 @@ def test_gradient_clipping_algorithm_with_deepspeed_enabled(
...
+
+    def _auto_wrap_policy(module: torch.nn.Module, recurse: bool, unwrapped_params: int) -> bool:
+        if recurse:
+            return True
+        if hasattr(module, '_fsdp_wrap'):
+            return bool(module._fsdp_wrap)
+        return False
```
Comment (mvpatel2000) [inline tests/algorithms/test_gradient_clipping.py:157]: nit: can we move this up outside fn?

File: tests/algorithms/test_gradient_clipping.py [hunk 3]
```diff
@@ -127,26 +134,44 @@ def test_gradient_clipping_algorithm_with_deepspeed_enabled(
...
+
+    device = DeviceGPU()
+    if not dist.is_initialized():
+        dist.initialize_dist(device=device, timeout=300)
```
Comment (mvpatel2000) [inline tests/algorithms/test_gradient_clipping.py:161]: handled by fixtures, not necessary

File: tests/algorithms/test_gradient_clipping.py [hunk 4]
```diff
@@ -127,26 +134,44 @@ def test_gradient_clipping_algorithm_with_deepspeed_enabled(
...
+
     clipping_threshold = 0.1191
-    apply_gc_fn = Mock()
-    monkeypatch.setattr(gc_module, 'apply_gradient_clipping', apply_gc_fn)
     state = dummy_state
+    model = simple_model_with_grads
+    model = FullyShardedDataParallel(model, auto_wrap_policy=_auto_wrap_policy, device_id=torch.cuda.current_device())
+    state.model = model
```
Comment (mvpatel2000) [inline tests/algorithms/test_gradient_clipping.py:167]: can we combine to one line

**Review by mvpatel2000 — APPROVED** [review]
LGTM
